src/library/clementine/clementinedbconnection.cpp

```diff
@@ -0,0 +1,184 @@
+    qDebug() << "Close Clementine database";
+    m_database.close();
+}
+
+void ClementineDbConnection::setTrackCollection(TrackCollectionManager* pTrackCollection) {
+    m_pTrackCollectionManager = pTrackCollection;
+}
+
+bool ClementineDbConnection::open(const QString& databaseFile) {
+    m_database = QSqlDatabase::addDatabase("QSQLITE", "CLEMENTINE_DB_CONNECTION");
+    m_database.setHostName("localhost");
+    m_database.setDatabaseName(databaseFile);
+    m_database.setConnectOptions("SQLITE_OPEN_READONLY");
+
+    //Open the database connection in this thread.
+    if (!m_database.open()) {
+        m_database.setConnectOptions(); // clear options
+        qWarning() << "Failed to open Clementine database." << m_database.lastError();
+        return false;
+    } else {
+        qDebug() << "Successful opened Clementine database";
+        return true;
+    }
+}
+
+QList<struct ClementineDbConnection::Playlist> ClementineDbConnection::getPlaylists() {
+    QList<struct ClementineDbConnection::Playlist> list;
+    struct ClementineDbConnection::Playlist playlist;
+
+    QSqlQuery query(m_database);
+    query.prepare("SELECT ROWID, playlists.name FROM playlists ORDER BY playlists.name");
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+        return {};
+    }
+    while (query.next()) {
+        playlist.playlistId = query.value(0).toString();
+        playlist.name = query.value(1).toString();
+        list.append(playlist);
+    }
+    return list;
+}
+
+QList<struct ClementineDbConnection::PlaylistEntry>
+ClementineDbConnection::getPlaylistEntries(int playlistId) {
+    PerformanceTimer time;
+    time.start();
+
+    QList<struct ClementineDbConnection::PlaylistEntry> list;
+    struct ClementineDbConnection::PlaylistEntry entry;
+
+    QSqlQuery query(m_database);
+    query.setForwardOnly(true); // Saves about 50% time
+
+    QString queryString;
+
+    queryString = QString(
+            "SELECT "
+            "ROWID, "
+            "playlist_items.title, "    // 1
+            "playlist_items.filename, " // 2
+            "playlist_items.length, "   // 3
+            "playlist_items.artist, "   // 4
+            "playlist_items.year, "     // 5
+            "playlist_items.album, "    // 6
+            "playlist_items.rating, "   // 7
+            "playlist_items.genre, "    // 8
+            "playlist_items.track, "    // 9
+            //"playlist_items.DateAddedStamp, "   //
+            "playlist_items.bpm, "        // 10
+            "playlist_items.bitrate, "    // 11
+            "playlist_items.comment, "    // 12
+            "playlist_items.playcount, "  // 13
+            "playlist_items.composer, "   // 14
+            "playlist_items.grouping, "   // 15
+            "playlist_items.type, "       // 16
+            "playlist_items.albumartist " // 17
+            "FROM playlist_items "
+            "WHERE playlist_items.playlist = %1")
+                          .arg(playlistId);
```

Comment:
Use bindValue here: https://doc.qt.io/qt-5/qsqlquery.html#approaches-to-binding-values

Reply:
added in 77e628e